### DISCUSSION TITLE
Improve latency on reading data from robot.

### DISCUSSION THREAD

**Discussion by Gray-Stone** [Yaskawa-Global/YMConnect#24] [Q&A] [answered]
I would like to monitor the motion of the robot as it moves. The data I need is arm status (joint angle), and general status (e-stop/is-done-moving/hold) 

I tried to read them using
```
    recv_timer.Reset();
    PositionData current_js;
    status = motoman_controller->ControlGroup->ReadPositionData(ControlGroupId::R1, CoordinateType::Pulse, 0, 0, current_js);
    ThrowIfStates(status , "Getting Pulse");
    auto read_pose_dur = recv_timer.Lap<ch::microseconds>();

    // Do the Joint pulse convert
    PositionData converted_pos;
    status = motoman_controller->Kinematics->ConvertPosition(
        ControlGroupId::R1, current_js, KinematicConversions::PulseToJointAngle, converted_pos);
    ThrowIfStates(status, "Converting pulse to deg");
    auto conv_dur = recv_timer.Lap<ch::microseconds>();

    ControllerStateData status_data;
    ThrowIfStates(motoman_controller->Status->ReadState(status_data) , "read status");
    auto status_dur = recv_timer.Lap<ch::microseconds>();

```

Where each access to `motoman_controller` takes about 10ms. Even the pulse to degree conversion. I assume the reason is that every operation is one round of TCP request/response, even with the conversion. 

In general, the combined 30ms latency is not too big a issue. My main concern is the duration stack up as I fetch more items, (say I also want to record TCP values along with joint values, this would add another 10ms to latency from conversion)

Are there ways to reduce this or combine reading all chunks of data in one pass so the latency/cycle time of fetching data is not getting longer with the amount of data I'm fetching? 

A side question, is it possible to add a timestamp when retrieving robot position data (controller time of when the joint values is captured), so the post calculation of velocity and acceleration could be a lot more accurate?

**Comment by ted-miller** [comment, marked as answer]
Hi @Gray-Stone 
YMConnect is not capable of deterministic real-time response. The latency that you're experiencing is normal and cannot be improved. Additionally, there is no ability to add a controller timestamp using this interface.

For applications where latency is a concern, I recommend the [MotoPlus SDK](https://www.motoman.com/en-us/products/software/development/motoplus-sdk). It is more difficult to develop, but offers the best possible performance for real time control.

**Comment by vinaym815** [comment]
@ted-miller I am thinking of starting a subtask that formats all the necessary data into a string variable and reading it. 
I wanted to ask if there is limit on the string length ? I couldn't find any information about it in the manuals.
Thank you.

**Reply by ted-miller** [reply]
Hello @vinaym815 
Yes, String (S) variables are limited to 32 bytes. For single-byte character-sets, like English, you'll get the full 32 characters. But for multi-byte languages, such as Japanese, you'll be limited to 16 characters.

**Reply by vinaym815** [reply]
@ted-miller Thank you.

**Reply by vinaym815** [reply]
I ended up reducing the YMConnect calls by: 
1) editing the user ladder program to save to robot status and some extra variables to a register; and subsequently reading the register
2) Computing the joint angle to pulse ration from jogging.
